In [1]:
import nltk

In [2]:
nltk.download()

showing info https://raw.githubusercontent.com/nltk/nltk_data/gh-pages/index.xml


True

# Tokenization

## Sentence Tokenization

In [8]:
from nltk.tokenize import sent_tokenize
text="""Dogs ,Canis lupus familiaris, are domesticated mammals,
not natural wild animals. They were originally bred from
wolves. They have been bred by humans for a long time, and
were the first animals ever to be domesticated.
Today, some dogs are used as pets, others are used to help
humans do their work. They are a popular pet because they
are usually playful, friendly, loyal and listen to humans.
Thirty million dogs in the United States are registered as
pets. Dogs eat both meat and vegetables, often mixed
together and sold in stores as dog food. Dogs often have
jobs, including as police dogs, army dogs, assistance dogs,
fire dogs, messenger dogs, hunting dogs, herding dogs, or
rescue dogs.
They are sometimes called "canines" from the Latin word for dog
- canis. Sometimes people also use "dog" to describe other
canids, such as wolves. A baby dog is called a pup or puppy.
A dog is called a puppy until it is about one year old.
Dogs are sometimes referred to as "man’s best friend" because
they are kept as domestic pets and are usually loyal and
like being around humans."""
tokenized_text = sent_tokenize(text)
print(*tokenized_text, sep="\n\n")

Dogs ,Canis lupus familiaris, are domesticated mammals,
not natural wild animals.

They were originally bred from
wolves.

They have been bred by humans for a long time, and
were the first animals ever to be domesticated.

Today, some dogs are used as pets, others are used to help
humans do their work.

They are a popular pet because they
are usually playful, friendly, loyal and listen to humans.

Thirty million dogs in the United States are registered as
pets.

Dogs eat both meat and vegetables, often mixed
together and sold in stores as dog food.

Dogs often have
jobs, including as police dogs, army dogs, assistance dogs,
fire dogs, messenger dogs, hunting dogs, herding dogs, or
rescue dogs.

They are sometimes called "canines" from the Latin word for dog
- canis.

Sometimes people also use "dog" to describe other
canids, such as wolves.

A baby dog is called a pup or puppy.

A dog is called a puppy until it is about one year old.

Dogs are sometimes referred to as "man’s best friend

In [9]:
len(tokenized_text)

13

There are 13 sentences, seperated by full stop character(".")

The Third sentence is "They have been bred by humans for a long time, and
were the first animals ever to be domesticated."

## Word Tokenization

In [11]:
from nltk.tokenize import word_tokenize
tokenized_word=word_tokenize(text)
print(*tokenized_word, sep="\n")

Dogs
,
Canis
lupus
familiaris
,
are
domesticated
mammals
,
not
natural
wild
animals
.
They
were
originally
bred
from
wolves
.
They
have
been
bred
by
humans
for
a
long
time
,
and
were
the
first
animals
ever
to
be
domesticated
.
Today
,
some
dogs
are
used
as
pets
,
others
are
used
to
help
humans
do
their
work
.
They
are
a
popular
pet
because
they
are
usually
playful
,
friendly
,
loyal
and
listen
to
humans
.
Thirty
million
dogs
in
the
United
States
are
registered
as
pets
.
Dogs
eat
both
meat
and
vegetables
,
often
mixed
together
and
sold
in
stores
as
dog
food
.
Dogs
often
have
jobs
,
including
as
police
dogs
,
army
dogs
,
assistance
dogs
,
fire
dogs
,
messenger
dogs
,
hunting
dogs
,
herding
dogs
,
or
rescue
dogs
.
They
are
sometimes
called
``
canines
''
from
the
Latin
word
for
dog
-
canis
.
Sometimes
people
also
use
``
dog
''
to
describe
other
canids
,
such
as
wolves
.
A
baby
dog
is
called
a
pup
or
puppy
.
A
dog
is
called
a
puppy
until
it
is
about
one
year
old
.
Dogs
are
sometimes
referre

In [12]:
len(tokenized_word)

229

In [13]:
len(word_tokenize(tokenized_text[0]))

15

In [15]:
len(word_tokenize(tokenized_text[-1]))

30

In [16]:
len(set(tokenized_word)) == len(tokenized_word)

False

There are 229 words in the paragraph, seperated by space " " filter LF out

There are 15 words in the first paragraph

There are 30 words in the last paragraph

There exists some words that duplicate

## Frequency Distribution

In [17]:
from nltk.probability import FreqDist
fdist = FreqDist(tokenized_word)
print(fdist)

<FreqDist with 120 samples and 229 outcomes>


In [18]:
fdist

FreqDist({',': 18, '.': 13, 'are': 10, 'dogs': 10, 'as': 7, 'and': 6, 'to': 5, 'dog': 5, 'Dogs': 4, 'They': 4, ...})

In [19]:
fdist.most_common(10)

[(',', 18),
 ('.', 13),
 ('are', 10),
 ('dogs', 10),
 ('as', 7),
 ('and', 6),
 ('to', 5),
 ('dog', 5),
 ('Dogs', 4),
 ('They', 4)]

## Stopwords

In [20]:
my_stopwords = ["is", "a", "this", ","]

from nltk.corpus import stopwords
nltk_stop_words=set(stopwords.words("english"))
print(nltk_stop_words)

{'me', 'being', 'shan', 'here', 'after', 'but', "you've", 'few', 'at', 's', 'does', 'herself', 'we', 'do', 'before', "wouldn't", 'between', 'you', "shouldn't", 'into', 'with', 'the', 'they', "shan't", 'yourselves', 'there', 'she', 'then', 'her', 'him', 've', 'by', 'of', 'll', 'needn', "isn't", 'why', 'theirs', 'yours', 'no', 'don', 'my', 'own', 'above', 'hadn', 'were', 'when', 'for', 'did', 'just', 'ain', 'this', 'are', 'itself', 'should', 'doesn', "won't", 'while', 'any', 'below', 'been', 'themselves', "you'll", 'on', 'during', 'has', 'further', 'if', 'only', 'ourselves', 'whom', "couldn't", 'that', 'through', 'again', 'both', 'didn', "mustn't", 'm', "didn't", 'our', 'some', 'himself', "mightn't", 'what', 'won', "wasn't", 'be', 'yourself', 'am', "hasn't", 'wouldn', 'was', 'is', "it's", 'as', 'about', 'out', 'o', "should've", 'mustn', 'have', 'aren', 'his', 'shouldn', 'an', 'will', 'their', 'which', 'than', 'until', "aren't", 'mightn', "you're", 'off', 'it', 'now', 'up', 'over', 're', 

stopwords can be ommited in many sentences, they are somewhat unnecessary for interpret the meaning of the sentence

## Removing Stopwords

In [26]:
filtered_words=[]
for w in tokenized_word:
    if w not in my_stopwords:
        filtered_words.append(w)
print("Tokenized Sentence:",tokenized_word, len(tokenized_word))
print("\n")
print("Filterd Sentence:",filtered_words, len(filtered_words))


Tokenized Sentence: ['Dogs', ',', 'Canis', 'lupus', 'familiaris', ',', 'are', 'domesticated', 'mammals', ',', 'not', 'natural', 'wild', 'animals', '.', 'They', 'were', 'originally', 'bred', 'from', 'wolves', '.', 'They', 'have', 'been', 'bred', 'by', 'humans', 'for', 'a', 'long', 'time', ',', 'and', 'were', 'the', 'first', 'animals', 'ever', 'to', 'be', 'domesticated', '.', 'Today', ',', 'some', 'dogs', 'are', 'used', 'as', 'pets', ',', 'others', 'are', 'used', 'to', 'help', 'humans', 'do', 'their', 'work', '.', 'They', 'are', 'a', 'popular', 'pet', 'because', 'they', 'are', 'usually', 'playful', ',', 'friendly', ',', 'loyal', 'and', 'listen', 'to', 'humans', '.', 'Thirty', 'million', 'dogs', 'in', 'the', 'United', 'States', 'are', 'registered', 'as', 'pets', '.', 'Dogs', 'eat', 'both', 'meat', 'and', 'vegetables', ',', 'often', 'mixed', 'together', 'and', 'sold', 'in', 'stores', 'as', 'dog', 'food', '.', 'Dogs', 'often', 'have', 'jobs', ',', 'including', 'as', 'police', 'dogs', ',', '

In [27]:
filtered_words=[]
for w in tokenized_word:
    if w not in stopwords.words("english"):
        filtered_words.append(w)
print("Tokenized Sentence:",tokenized_word, len(tokenized_word))
print("\n")
print("Filterd Sentence:",filtered_words, len(filtered_words))

Tokenized Sentence: ['Dogs', ',', 'Canis', 'lupus', 'familiaris', ',', 'are', 'domesticated', 'mammals', ',', 'not', 'natural', 'wild', 'animals', '.', 'They', 'were', 'originally', 'bred', 'from', 'wolves', '.', 'They', 'have', 'been', 'bred', 'by', 'humans', 'for', 'a', 'long', 'time', ',', 'and', 'were', 'the', 'first', 'animals', 'ever', 'to', 'be', 'domesticated', '.', 'Today', ',', 'some', 'dogs', 'are', 'used', 'as', 'pets', ',', 'others', 'are', 'used', 'to', 'help', 'humans', 'do', 'their', 'work', '.', 'They', 'are', 'a', 'popular', 'pet', 'because', 'they', 'are', 'usually', 'playful', ',', 'friendly', ',', 'loyal', 'and', 'listen', 'to', 'humans', '.', 'Thirty', 'million', 'dogs', 'in', 'the', 'United', 'States', 'are', 'registered', 'as', 'pets', '.', 'Dogs', 'eat', 'both', 'meat', 'and', 'vegetables', ',', 'often', 'mixed', 'together', 'and', 'sold', 'in', 'stores', 'as', 'dog', 'food', '.', 'Dogs', 'often', 'have', 'jobs', ',', 'including', 'as', 'police', 'dogs', ',', '

# Lexicon Normalization

## Stemming

In [34]:
from nltk.stem import PorterStemmer
stem = PorterStemmer()
word = "connection"
stem.stem(word)

'connect'

In [35]:
stem = PorterStemmer()
stem.stem("connectivity")

'connect'

In [36]:
stem = PorterStemmer()
stem.stem("connecting")

'connect'

In [37]:
stem = PorterStemmer()
stem.stem("derivational")

'deriv'

In [38]:
stem = PorterStemmer()
stem.stem("multiplying")

'multipli'

In [39]:
stem = PorterStemmer()
stem.stem("multiplicative")

'multipl'

Output of stem function is reduced word(-ish) of that word. By remove suffix such as "ing", "ly"

In [41]:
stem = PorterStemmer()
stemmed_words=[]
for w in filtered_words:
    stemmed_words.append(stem.stem(w))
print("Filtered Words:",filtered_words)
print("\n")
print("Stemmed Words:",stemmed_words)

Filtered Words: ['Dogs', ',', 'Canis', 'lupus', 'familiaris', ',', 'domesticated', 'mammals', ',', 'natural', 'wild', 'animals', '.', 'They', 'originally', 'bred', 'wolves', '.', 'They', 'bred', 'humans', 'long', 'time', ',', 'first', 'animals', 'ever', 'domesticated', '.', 'Today', ',', 'dogs', 'used', 'pets', ',', 'others', 'used', 'help', 'humans', 'work', '.', 'They', 'popular', 'pet', 'usually', 'playful', ',', 'friendly', ',', 'loyal', 'listen', 'humans', '.', 'Thirty', 'million', 'dogs', 'United', 'States', 'registered', 'pets', '.', 'Dogs', 'eat', 'meat', 'vegetables', ',', 'often', 'mixed', 'together', 'sold', 'stores', 'dog', 'food', '.', 'Dogs', 'often', 'jobs', ',', 'including', 'police', 'dogs', ',', 'army', 'dogs', ',', 'assistance', 'dogs', ',', 'fire', 'dogs', ',', 'messenger', 'dogs', ',', 'hunting', 'dogs', ',', 'herding', 'dogs', ',', 'rescue', 'dogs', '.', 'They', 'sometimes', 'called', '``', 'canines', "''", 'Latin', 'word', 'dog', '-', 'canis', '.', 'Sometimes', '

every words in the sentence reduced to the shortest form

## Lemmatization

In [43]:
from nltk.stem.wordnet import WordNetLemmatizer
lem = WordNetLemmatizer()
word = "connecting"
lem.lemmatize(word, pos="v")

'connect'

In [44]:
lem = WordNetLemmatizer()
word = "flying"
print(lem.lemmatize(word, pos="v"))
print(lem.lemmatize(word, pos="n"))

fly
flying


seem pos="v" change inputed word to verb form, and pos="n" change to noun form

## POS Tagging

In [46]:
sent = "I am studying at ICT"
tokens=nltk.word_tokenize(sent)
print(tokens)
nltk.pos_tag(tokens)

['I', 'am', 'studying', 'at', 'ICT']


[('I', 'PRP'),
 ('am', 'VBP'),
 ('studying', 'VBG'),
 ('at', 'IN'),
 ('ICT', 'NNP')]

The function assign part-of-speech type to each word

PRP -> personal pronoun

VBP -> verb, singular present, non-3rd person

IN -> preposition/subordinating conjunction

NNP -> proper noun, singular

etc.